src/EventStore.Core/TransactionLog/Chunks/TFChunkWriter.cs

```diff
@@ -40,24 +43,50 @@ public class TFChunkWriter : ITransactionFileWriter {
 			// DO NOTHING
 		}
 
+		public void Begin() {
+			if (_inTransaction)
+				throw new Exception("Attempt to begin a new transaction while there is an ongoing transaction.");
+
+			_inTransaction = true;
+		}
+
 		public bool Write(ILogRecord record, out long newPos) {
 			var result = _currentChunk.TryAppend(record);
 			if (result.Success)
-				_writerCheckpoint.Write(result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition);
+				_nextRecordPosition = result.NewPosition + _currentChunk.ChunkHeader.ChunkStartPosition;
 			else
-				CompleteChunk(); // complete updates checkpoint internally
-			newPos = _writerCheckpoint.ReadNonFlushed();
+				CompleteChunk();
+
+			if (!_inTransaction)
+				CommitInternal();
+
+			newPos = _nextRecordPosition;
 			return result.Success;
 		}
 
+		public void Commit() {
+			if (!_inTransaction)
+				throw new Exception("Attempt to commit while there is no ongoing transaction.");
+
+			_inTransaction = false;
+			CommitInternal();
+		}
+
+		private void CommitInternal(bool flush = false) {
+			_writerCheckpoint.Write(_nextRecordPosition);
+			if (flush)
+				_writerCheckpoint.Flush();
+		}
+
 		public void CompleteChunk() {
 			var chunk = _currentChunk;
 			_currentChunk = null; // in case creation of new chunk fails, we shouldn't use completed chunk for write
 
 			chunk.Complete();
```

Comment:
is there a problem in the end with the chunk becomming readonly with the writercheckpoint still in it (in case we end up not advancing the writer checkpoint)?

Reply:
if we end up not advancing the writer checkpoint, it won't be a problem because on startup, the last chunk is opened as an ongoing chunk file which is not read-only (even if the chunk already has a footer)